man/didump.8
.\" $Revision: 1.11 $
.TH DIDUMP 8
.TH DILOAD 8
.SH NAME
didump \- Diablo program to dump the dhistory database
.PP
diload \- Diablo program to append to the dhistory database
.SH SYNOPSIS
.B didump
[
.B \-e
]
[
.B \-f
]
[
.B \-H
]
[
.B \-h
]
[
.B \-l
]
[
.B \-n
]
[
.B \-o
]
[
.B \-p
]
[
.B \-rN
]
[
.B \-TN
]
[
.B \-t
]
[
.B \-v
]
[
.B \-x
]
[
.B \-C diablo.config
]
[
.B \-d[n]
]
[
.B \-V
]
[
.B \@offset
]
.B dhistory-file
.PP
.B diload
[
.B \-f
]
[
.B \-h hashtablesize
]
[
.B \-u
]
[
.B \-v
]
[
.B \-C diablo.config
]
[
.B \-d[n]
]
[
.B \-V
]
[
.B dhistory-file
]

.SH DESCRIPTION
.IR Didump
dumps the specified dhistory file into a human readable format.  It is 
used to debug the dhistory database, to recover a corrupted dhistory 
file.  In combination with diload, didump may be used to remove expired
entries over rememberdays days old from the dhistory file.
.PP
.B -e
is used to cause didump to generate ENTRIES lines - useful for diload
progress.
.PP
.B -f
with a normal dump (no -t), seeks to the end of the history file and
prints out entries as other programs (diablo) add them.  DiDump will
not exit in this case.  This option is useful to get an idea of the
aggregate incoming article rate.  NOTE, however, that diablo does not
add history file entries with a simple append, so you may see blank records
in the output that, in fact, are filled-in later.
.PP
.B -H
Dump a hash trace for specified Message-ID or hash.
.PP
.B -h
For each entry dumped, lookup the hash in the history file to verify
that it is available via normal lookup.
.PP
.B -l
is used to force didump to flush each line to the output rather then buffer
the output, useful when running "didump -f dhistory | dreadart -v -f -"
or something of that ilk.
.PP
.B -n
Also print the value of the 'next' pointer.
.PP
.B -o
didump V1.08 can read older (V1.07 or smaller) Diablo dhistory files if this
option is specified.  This is used when upgrading to 1.08 to reload the
dhistory file with the dhistory file format.
.PP
.B -p
Displays progress on stdout. Note that the total records may not
be accurate if the -x or -T options are used because didump doesn't
know how many records will be displayed.
.PP
.B -rN
Set the rememberdays value to N. The rememberdays value specifies
the number of days worth of expired articles to keep. The default
is obtained from ``diablo.config'' and a value of '-1' will cause
didump to not dump any expired articles. If the value has 's', 'm'
or 'h' appended, the numeric value is taken as seconds, minutes or
hours, respectively. This option has no effect without the ``-x'' option.
.PP
.B -TN
Don't dump records older than N seconds. The time value can be
suffixed with 'd', 'h' or 'm' to specify days, hours or minutes.
The default is '-1' which means dump all records.
.PP
.B -t
is used for debugging and causes didump to attempt to trace the dhistory
file's hash table links.  It is very inefficient.
.PP
.B -v
is used to generate additional information, such as the spool file 
containing the article.
.PP
.B -x
causes didump to not dump expired entries that are over rememberdays days old.
.PP
.B -C
allows the specification of the path of the diablo.config file to use.
.PP
.B -d[n]
will set debug mode, at optional level n.
.PP
.B -V
will print the version and exit.
.PP
Generally speaking, when recovering a dhistory file or trimming off 
obsolete entries,
.B didump
or 
.B didump -x
is run with the existing history file as an argument, and the output is
piped to
.IR diload 
\-f tempfile.
Once the recovery is complete, the temporary history file is then renamed
to become the actual history file.  The diablo server must be paused
during such operations to prevent it from acting on the 'old' version
of the history file while the new one is being constructed.

.PP
.IR Diload
takes history data on the standard input and appends it to the
specified history file.  If no history file is specified, the
active 
.B dhistory
database is used.  If any given record already exists in the dhistory file, 
the new record is thrown away and the original record left intact.  You can
safely run
.IR diload
to append new entries without shutting the main Diablo server down, but
Diablo must be properly paused and restarted if you intend to rewrite the history file.
.PP
.B -f
Turn on special-fast mode.  This locks the history file through the
entire procedure and makes the assumption that entries being added
did not previously exist.  This also causes the base hash table to
be cached and written then written out when the load operation has
completed.  DiLoad will create the history file you supply as an argument 
and will refuse to run if it is already pre-existing.  This
is used in the weekly trimming script to quickly rebuild the history
file from a dump.
.PP
.B -h hashtablesize
The default hash table size is configured in the diablo.config
configuration file, and is the best way to set the size.  Alternatively,
you can use this option to set it to any power of two.  For a transit
server, the suggested size is 4m or 8m.  For a spool server, the optimal
size is a power of two slightly less than the number of articles you plan
to store (i.e. if you have 90 million articles, 64m or perhaps 32m).  Do 
NOT forget the 'm' for million, or you will be a very unhappy camper.
Please note that if you create a larger history hash in this manner, the 
next run of diload (perhaps from biweekly.atrim) will still only create
a table the size of whatever the configured default is, so be sure to
modify your diablo.config or biweekly.atrim to reflect any appropriate
changes.
.PP
.B -u
May be used with special-fast mode to re-enable checking for duplicates.
You have the advantage of locking the history file throughout the
entire operation, but can still use a pre-existing history file to 'update'.
.PP
.B -v
will turn on verbose mode.  diload prints out its progress.
.PP
.B -C
allows the specification of the path of the diablo.config file to use.
.PP
.B -d[n]
will set debug mode, at optional level n.
.PP
.B -V
will print the version and exit.

.SH "SEE ALSO"
diablo(8), 
dicmd(8),
didump(8),
diload(8),
dnewslink(8),
doutq(8),
dexpire(8),
dexpireover(8),
diconvhist(8),
dilookup(8),
dspoolout(8),
dnewsfeeds(5),
dkp(8),
dpath(8),
diablo-kp(5),
diablo-files(5)
.PP



RELEASE_NOTES

			    RELEASE NOTES

		  DIABLO NEWS TRANSIT AND READER SYSTEM

		      diablo-users@openusenet.org
		    http://www.openusenet.org/diablo/
			   news.software.nntp

Known Issues:
	- OLD: dreaderd: problem with corrupt header files caused
	  by what seems to be a lock race condition with 'dexpireover -R'.
	  This problem has existed for years.

Experimental (features/utilities that are new and/or not properly tested
  	      and hence should be used with care on production systems):
	- Compressed spools
	- Compressed feeds
	- IPv6 support
	- dnewsfeeds %include option
	- dexpirecache
	- dcancel

Summary of changes likely to impact an upgrade from 5.x
	- new dreader header index format. Cannot go back to
	  older versions as they will not be able to read
	  the new version.

Summary of changes likely to impact an upgrade from 4.x
	- history version 2 (offsets converted to indexes)
	- new history expire (dhisexpire)
	- fixed bug in dexpire.ctl inheritance handling

Summary of changes likely to impact an upgrade from 3.x
	- New spool storage format. Each article is stored with a binary
	  header. This new format is not usable with older versions of
	  diablo, although 4.x can read prior storage formats.

Diablo 5-current

2009-05-25
	* dreaderd: Don't know why no one ever noticed that tcp
	  socket options per user weren't getting set right.

2009-04-20
        * dcancel: Add the ability to cancel articles by article
          number in the reader header index. (from Michael Dorrington)

2009-04-18
	* Oops, never committed the stunnel patch for useproxied.

2009-02-01
	* parsedate wouldn't deal well with certain days of the
	  week that weren't capitalized

2009-01-28
	* Cleanup of various code problems, including new expire
	  cleanup, spaces-in-Newsgroups cleanup, etc.

2009-01-15
	* Implement alternate article numbering views for clients
	  (this version still needs some cleanup).  Clients may
	  indicate their capabilities via the XNUMBERING extension.

	  RFC977 imposes no real limits on article numbering.  This
	  may cause some issues with non-64-bit-clean clients.

	  RFC3977 imposes a 2^31-1 limit AND places restrictions
	  on the resetting of the low-water mark, which is excessively
	  retarded because it effectively means that once a group hits
	  the limit, it's just dead, but it's the behaviour that is
	  defined in the RFC.

	  WINDOW guarantees compatibility with non-64-bit-clean
	  clients by restricting the article numbers to a window
	  between 1 and 2^31-1; further, the number of articles in
	  a group is limited to guarantee that most clients should
	  detect it as a "group reset" along the various lines of
	  ill-defined logic often used by clients.  It is not a
	  frequent event, but when it happens, this has the best
	  possibility of being successful without being too
	  disruptive.

	  Another patch shortly will clean this code up, and will
	  allow for selection of which view to present by default
	  to clients who do not explicitly request a particular
	  view.

2008-12-24
	* Absolutely huge changes that have been underway for more
	  than half a year.  

	  There are a set of interrelated but different problems
	  going on in the system.

	  The first is that the legacy overview code was not 64-bit
	  safe; this effectively limited overviews for a group to
	  around 60 million articles and then wackiness happened.  
	  The code has been walked through and updated so that on 
	  64-bit platforms, larger overviews are possible (we've 
	  exercised it to 500 mil). 32-bit platforms currently lose
	  due to the inability of the operating system to map large 
	  files, but support for this is also possible and a cleanly 
	  coded solution may be committed in the future.

	  Further, article numbering is a disaster waiting to happen.
	  The legacy code is organized to handle 31-bit message 
	  numbers.  Certain alt.binaries high volume groups are 
	  poised to pass message number 2^31-1.  Now, on one hand,
	  the problem could be solved by simply 64-bit-ifying the
	  code to handle 63-bit article numbers.  On the other hand,
	  some nimrods decided to specify 2^31-1 as a maximum article
	  number in RFC3977, which, fortunately, we don't really 
	  claim compliance with.  This code revision simply extends
	  the article numbering to 2^63.  In doing so, it will update
	  your overview files from VERSION 3 to VERSION 4.  This is
	  a harmless change except for groups where article numbering
	  exceeds 2^31-1.  Some effort was put into keeping the files
	  otherwise-compatible with older code.

	  Note that if you are using Diablo as an article numbering
	  master, you will need to update Diablo AND dreaderd.

	  People running this codebase will be able to have larger
	  overviews and won't run into a problem at artno 2^31-1.

2008-12-22
	* Various compiler warning cleanups

2008-12-19
	* Runaway spool thread status flag repurposing, based on a 
	  strategy POC'd by tim@euroaccess (thanks for the headaches, 
	  grin)

2008-12-18
	* Various compiler warning cleanups

2008-12-18
	* It's 2008.  Everyone has zlib.  If I don't hear objections, I
	  will probably be slowly removing conditional support for it.

2008-12-18
	* Add error checking for group resizing; many platforms limit
	  calloc memory availability and application of an "unlimit" or
	  equivalent in the daily.reader script might be required.  This
	  just adds a reasonable error message

2008-12-18
	* Add compressed header transmission to clients (XZVER support)
	  Newshosting and a few other places have been running this for
	  almost a year without any complaints, but it is still in
	  testing.

2008-12-16
	* drgpctl masscancel: support cancellation of large numbers of
	  articles from very large groups (one-pass through the overview
	  file)

2008-12-15
	* Dump threads status from sigUsr1 (can't find who submitted this
	  patch, apologies, whoever)

2008-12-15
	* Report virtual server name in log lines


util/didump.c

/*
 * DIDUMP.C	Dump or trace a dhistory file
 *
 * (c)Copyright 1997, Matthew Dillon, All Rights Reserved.  Refer to
 *    the COPYRIGHT file in the base directory of this distribution 
 *    for specific rights granted.
 */

#include "defs.h"

/*
 * OldHistory - versions <= 1.07
 *
 */

typedef struct OldHistory {
    uint32	next;   /* next link            */
    uint32	gmt;    /* gmt time in minutes  */
    hash_t      hv;     /* hash value           */
    uint16	iter;   /* file id              */
    uint16	exp;    /* hours relative to gmt minutes */
} OldHistory;

int TraceMode = 0;
int DNPOpt = 0;
int FOpt = 0;
int FCount = 0;
int OldOpt = 0;
int StartOff = 0;
int LineModeOpt = 0;
int VerboseOpt = 0;
int QuietOpt = 0;
int VerifyHistory = 0;
int NextOpt = 0;
int EntriesOpt = 0;
int EmptyOpt = 0;
int ShowProgress = 0;
int DumpHashOnly = 0;
char *FindHash = NULL;
time_t MaxAge = -1;
int HistoryVersion = 0;
uint32 HOffset = 0;

uint32 ExpireDropCount = 0;
uint32 ExpireKeepCount = 0;
uint32 MaxAgeCount = 0;
uint32 ZeroGmtCount = 0;
uint32 LookupCount = 0;
uint32 DumpedCount = 0;

void DumpTrace(int fd, int hsize, int rsize);
void DumpQuick(int fd, int hsize, int rsize);
void DumpChain(int fd, int hsize, int rsize, hash_t *hv);

void
Usage(void)
{
    printf("Dump the history file entries to stdout.\n\n");
    printf("didump [-e] [-f] [-H msgid|hash] [-h] [-l] [-n] [-o] [-p] [-r remember] [-t]\n");
    printf("\t[-TN] [-v] [-x] [-C diablo.config] [-d[n]] [-V] [@offset] dhistory-file\n");
    printf("  where:\n");
    printf("\tdefault\t- quick dump\n");
    printf("\t-e\t- also dump an ENTRIES line - useful for diload progress\n");
    printf("\t-f\t- 'tail -f' the history file\n");
    printf("\t-H\t- dump hash trace for specified Message-ID or hash\n");
    printf("\t-h\t- display entries not found with a HistoryLookup\n");
    printf("\t-l\t- line mode - flush output after every line\n");
    printf("\t-m\t- dump msgid hashes only\n");
    printf("\t-n\t- also dump the value of the 'next' pointer\n");
    printf("\t-o\t- dump old-style (diablo V <= 1.07) history file\n");
    printf("\t-p\t- show progress on stderr\n");
    printf("\t-q\t- quiet - don't show stats\n");
    printf("\t-rN\t- set rememberdays\n");
    printf("\t-t\t- hash table trace (slow)\n");
    printf("\t-TN\t- don't dump articles older than N seconds\n");
    printf("\t-v\t- Include additional (synthesized) info\n");
    printf("\t-x\t- do not dump records older than rememberdays old\n");
    printf("\t-z\t- include entries with a gmt of zero in the dump\n");
    printf("\t-Cfile\t- specify diablo.config to use\n");
    printf("\t-d[n]\t- set debug [with optional level]\n");
    printf("\t-V\t- print version and exit\n");
    printf("\t@\t- specify the starting offset (in bytes)\n");
    exit(1);
}

int
main(int ac, char **av)
{
    int fd;
    int i;
    int hsize = 1024 * 1024;
    int rsize = sizeof(History);
    struct stat st;
    char *fileName = NULL;

    LoadDiabloConfig(ac, av);

    for (i = 1; i < ac; ++i) {
	char *ptr = av[i];

	if (*ptr != '-') {
	    if (*ptr == '@') {
		StartOff = strtol(ptr + 1, NULL, 0);
		continue;
	    }
	    if (fileName) {
		fprintf(stderr, "unexpected argument\n");
		exit(1);
	    }
	    fileName = ptr;
	    continue;
	}
	ptr += 2;
	switch(ptr[-1]) {
	case 'e':
	    EntriesOpt = 1;
	    break;
	case 'f':
	    FOpt = 1;
	    if (*ptr)
		FCount = strtol(ptr, NULL, 0);
	    break;
	case 'H':
	    FindHash = (*ptr ? ptr : av[++i]);
	    break;
	case 'h':
	    VerifyHistory = 1;
	    break;
	case 'l':
	    LineModeOpt = 1;
	    break;
	case 'm':
	    DumpHashOnly = 1;
	    break;
	case 'n':
	    NextOpt = 1;
	    break;
	case 'o':
	    OldOpt = 1;
	    break;
	case 'p':
	    ShowProgress = 1;
	    break;
	case 'q':
	    QuietOpt = 1;
	    break;
	case 'r':
	    if (!*ptr)
		ptr = av[++i];
	    DOpts.RememberSecs = TimeSpec(ptr, "d");
	    if (DOpts.RememberSecs == -1)
		Usage();
	    break;
	case 'T':
	    MaxAge = btimetol(*ptr ? ptr : av[++i]);
	    break;
	case 't':
	    TraceMode = 1;
	    break;
	case 'v':
	    if (*ptr)
		VerboseOpt = strtol(ptr, NULL, 0);
	    else
		++VerboseOpt;
	    break;
	case 'x':
	    DNPOpt = 1;
	    break;
	case 'z':
	    EmptyOpt = 1;
	    break;
	/* Common options */
	case 'C':		/* parsed by LoadDiabloConfig */
	    if (*ptr == 0)
		++i;
	    break;
	case 'd':
	    DebugOpt = 1;
	    if (*ptr)
		DebugOpt = strtol(ptr, NULL, 0);
	    break;
	case 'V':
	    PrintVersion();
	    break;
	default:
	    if (isdigit((int)ptr[-1])) {
		FCount = strtol(ptr - 1, NULL, 0);
	    } else {
		fprintf(stderr, "illegal option: %s\n", ptr - 2);
		Usage();
	    }
	}
    }

    if (fileName == NULL) {
	Usage();
    }

    if (VerifyHistory || FindHash != NULL)
	HistoryOpen(fileName, HGF_READONLY);

    if (VerboseOpt)
	LoadSpoolCtl(0, 1);


    if (OldOpt) {
	char *paramName = malloc(strlen(fileName) + 32);
	FILE *fi;

	sprintf(paramName, "%s.param", fileName);

	rsize = sizeof(OldHistory);

	if ((fi = fopen(paramName, "r")) != NULL) {
	    char buf[256];

	    hsize = 0;

	    while (fgets(buf, sizeof(buf), fi) != NULL) {
		if (buf[0] == 'H')
		    hsize = strtol(buf + 1, NULL, 0);
	    }
	    fclose(fi);

	    if (hsize == 0) {
		fprintf(stderr, "dhistory parameter file error\n");
		exit(1);
	    }
	}
    }

    if (DNPOpt && TraceMode) {
	fprintf(stderr, "-x only works for quick dumps\n");
	exit(1);
    }

    if ((fd = open(fileName, O_RDONLY)) >= 0 && fstat(fd, &st) == 0) {
	/*
	 * new style history file has a header
	 */

	if (OldOpt == 0) {
	    HistHead hh;

	    if (read(fd, &hh, sizeof(hh)) != sizeof(hh)) {
		fprintf(stderr, "corrupted history file\n");
		exit(1);
	    }
	    if (hh.hmagic != HMAGIC) {
		fprintf(stderr, "corrupted history file\n");
		exit(1);
	    }
	    if (hh.version > HVERSION) {
		fprintf(stderr, "WARNING! Version mismatch file V%d, expecting V%d\n", hh.version, HVERSION);
		fprintf(stderr, "dump may be invalid\n");
	    }
	    HistoryVersion = hh.version;
	    rsize = hh.henSize;
	    hsize = hh.hashSize;
	    HOffset = hh.headSize + hsize * sizeof(HistIndex);

	    lseek(fd, hh.headSize, 0);
	}

	if (!QuietOpt)
	    fprintf(stderr, "Dumping, hash table %d entries, record size %d\n",
							hsize, rsize);

	if (TraceMode) {
	    DumpTrace(fd, hsize, rsize);
	} else if (FindHash != NULL) {
	    hash_t hv;
	    if (*FindHash == '<')
		hv = hhash(FindHash);
	    else
		sscanf(FindHash, "%x.%x", &hv.h1, &hv.h2);
	    DumpChain(fd, hsize, rsize, &hv);
	} else {
	    DumpQuick(fd, hsize, rsize);
	}
	close(fd);
    } else {
	fprintf(stderr, "open failed\n");
	exit(1);
    }
    return(0);
}

void
PrintTrace(int fd, HistIndex index, int rsize)
{
    off_t off;
    int maxChainLen = 1000;

    while (index) {
	History h = { 0 };

	if (HistoryVersion > 1)
	    off = (off_t)HOffset + (off_t)index * sizeof(History);
	else
	    off = index;
	lseek(fd, off, 0);
	if (read(fd, &h, rsize) != rsize) {
	    fprintf(stderr, "read error @ %d (%lld)", index, (long long)off);
	    break;
	} else {
	    printf(" [%u,%u %08x.%08x.%04x gm=%d ex=%d boff=%d bsize=%d F=%s part=%d]",
		index, (unsigned int)h.next,
		h.hv.h1,
		h.hv.h2,
		(int)h.iter,
		(int)h.gmt,
		(int)h.exp,
		(int)h.boffset,
		(int)h.bsize,
		((h.exp & EXPF_HEADONLY) ? "H" : ""),
		(int)H_SPOOL(h.exp)
	    );
	}
	if (VerboseOpt) {
	    char buf[1024];
	    printf("\n");
	    ArticleFileName(buf, sizeof(buf), &h, ARTFILE_FILE);
	    printf("\tfile=\"%s\"", buf);
	    {
		struct tm *tp;
		time_t t;
		char tbuf[64];

		t = h.gmt * 60;   
		tp = localtime(&t);
		strftime(tbuf, sizeof(tbuf), "%d-%b-%Y %H:%M:%S", tp);
		printf("\ttime=%s\n", tbuf);
	    }
	}
	index = h.next;
	if (--maxChainLen == 0) {
	    printf(" MAXIMUM CHAIN LENGTH EXCEEDED!");
	    break;
	}
    }
    if (index)
	printf(" offset error: %d", index);
    printf("\n");
}

void
DumpTrace(int fd, int hsize, int rsize)
{
    int i;
    HistIndex *Ary = calloc(hsize, sizeof(HistIndex));

    if (read(fd, Ary, hsize * sizeof(HistIndex)) != hsize * sizeof(HistIndex)) {
	fprintf(stderr, "Unable to read hash table array\n");
	exit(1);
    }

    for (i = 0; i < hsize; ++i) {
	if (Ary[i] != 0) {
	    printf("Index %d: ", i);
	    PrintTrace(fd, Ary[i], rsize);
	}
    }
}

void
DumpChain(int fd, int hsize, int rsize, hash_t *hv)
{
    uint32 *Ary = calloc(hsize, sizeof(uint32));
    uint32 off;

    if (read(fd, Ary, hsize * sizeof(uint32)) != hsize * sizeof(uint32)) {
	fprintf(stderr, "Unable to read hash table array\n");
	exit(1);
    }

    off = Ary[(hv->h1 ^ hv->h2) & (hsize - 1)];
    PrintTrace(fd, off, rsize);
}

void
DumpQuick(int fd, int hsize, int rsize)
{
    char *hbuf;
    int n;
    int hlen;
    int rememberMins = DOpts.RememberSecs / 60;
    uint32 gmt = time(NULL) / 60;
    History th;
    off_t seekpos = 0;
    uint32 totalentries = 0;
    uint32 count = 0;
    int i;
    History *h;


    hlen = rsize * 4096;;
    if ((hbuf = (char *)malloc(hlen)) == NULL) {
	fprintf(stderr, "ERROR: Unable to malloc %dB for history cache (%s)\n",
				hlen, strerror(errno));
	return;
    }
    if (StartOff)
	seekpos = lseek(fd, StartOff, 0);
    else if (FOpt || FCount)
	seekpos = lseek(fd, -FCount * rsize, 2);
    else
	if (HistoryVersion > 1)
	    seekpos = lseek(fd, (off_t)hsize * sizeof(HistIndex) + (off_t)rsize, 1);
	else
	    seekpos = lseek(fd, (off_t)hsize * sizeof(HistIndex), 1);

    {
	struct stat st;

	if (fstat(fd, &st) == -1) {
	    fprintf(stderr, "Unable to fstat history: %s\n", strerror(errno));
	    exit(1);
	}
	totalentries = (int)(((double)st.st_size - (double)seekpos) / (double)rsize);
	if (EntriesOpt)
	    printf("ENTRIES %u\n", totalentries);
    }

    if (!QuietOpt)
	fprintf(stderr, "@%lld  %u records (%d bytes per record)\n",
				(long long)lseek(fd, 0L, 1),
				totalentries, rsize);

top:
    while ((n = read(fd, hbuf, hlen)) > 0) {

	n /= rsize;
	gmt = time(NULL) / 60;

	for (i = 0; i < n; ++i) {
	    h = (History *)(hbuf + i * rsize);


	    if (ShowProgress && (count++ % 32768) == 0)
		fprintf(stderr, "%u/%u  (%d%%)    \r", count, totalentries,
			(int)((double)count * 100.0 / (double)totalentries));

	    /*
	     * A gmt of zero is an empty entry - don't print it out unless
	     * we are verifying history
	     */
	    if (h->gmt == 0 && !VerifyHistory && !EmptyOpt) {
		ZeroGmtCount++;
		continue;
	    }

	    /*
	     * If we specified a maximum age, then deal with it
	     */
	    if (MaxAge != -1) {
		int32 dgmt = (gmt - h->gmt) * 60;	/* Delta seconds */
		if (dgmt > MaxAge) {
		    MaxAgeCount++;
		    continue;
		}
	    }

	    if (DNPOpt && H_EXPIRED(h->exp)) {
		int32 dgmt = gmt - h->gmt;	/* DELTA MINUTES */
		if (dgmt < -rememberMins || dgmt > rememberMins) {
		    ExpireDropCount++;
		    continue;
		} else {
		    ExpireKeepCount++;
		}
	    }

	    if (VerifyHistory && HistoryLookupByHash(h->hv, &th) == 0) {
		LookupCount++;
		continue;
	    }

	    if (DumpHashOnly) {
		printf("%08x.%08x\n", h->hv.h1, h->hv.h2);
		continue;
	    }
	    printf("DUMP %08x.%08x.%04x gm=%d ex=%-2d",
		h->hv.h1,
		h->hv.h2,
		(int)h->iter,
		(int)h->gmt,
		(int)h->exp
	    );
	    if (!OldOpt) {
		printf(" boff=%-7d bsize=%-6d", (int)h->boffset, (int)h->bsize);
		if (VerboseOpt) {
		    char buf[1024];
		    ArticleFileName(buf, sizeof(buf), h, ARTFILE_FILE);
		    printf(" file=\"%s\"", buf);
		}
		printf(" flags=%s",
		    ((h->exp & EXPF_HEADONLY) ? "H" : "")
		);
		if (NextOpt)
		    printf(" next=%d", h->next);
	    }
	    printf("\n");
	    DumpedCount++;
	    if (LineModeOpt)
		fflush(stdout);
	}
    }
    if (n < 0)
	fprintf(stderr, "Error reading file: %s\n", strerror(errno));

    if (FOpt) {
	usleep(100000);
	goto top;
    }
    if (!QuietOpt && count > 0)
	fprintf(stderr, "%u/%u  (%d%%)    \n", count, totalentries,
			(int)((double)count * 100.0 / (double)totalentries));
    if (!QuietOpt)
	fprintf(stderr, "%u entries dumped\n", DumpedCount);
    if (VerifyHistory) {
	fprintf(stderr, "%u entries found\n", LookupCount);
    } else if (!QuietOpt) {
	fprintf(stderr, "%u expired entries kept\n", ExpireKeepCount);
	fprintf(stderr, "%u expired entries dropped\n", ExpireDropCount);
	fprintf(stderr, "%u dropped as beyond max age\n", MaxAgeCount);
	fprintf(stderr, "%u dropped with gmt=0\n", ZeroGmtCount);
    }
    if (!QuietOpt)
	printf(".\n");
    free(hbuf);
}

